### DISCUSSION TITLE
Stress testing Falcon with 10,000 connections

### DISCUSSION THREAD

**Discussion by edzhelyov** [ioquatix/ioquatix#10] [Q&A]
Hi Samuel,

I'm evaluating the performance and concurrency of Falcon serving a simple action. I'm stress testing with around 10,000 concurrent connections. I have the following app:
```ruby
require 'sinatra'

RESP = 'success'.freeze

get '/' do
  # sleep 1

  RESP
end
```
I test with various benchmark tools – ab, siege, wrk, ali, and custom script. In the custom script, I will open 10,000 connections from nodejs and then send a get request to receive the response.

I've 4 cores and 8 hyper-threads on my machine and when I run `falcon serve -b http://localhost:3000 -n 4` I'm able to establish 10,000 connections and successfully get the response for them. Running `ab -c 10000 -n 10000 http://localhost:3000/` receives 10,000 responses successfully.

When I run a single worker as `falcon serve -b http://localhost:3000 -n 1` I will get around 4,000 connections established from the full 10,000, the rest will error with `read ECONNRESET`. When I make a request only 1,250 will return a successful response. And when I test with only 1,250 connections I get 4-5 bad `ECONNRESET`, so I conclude that a single thread Falcon can handle around 1,2500 simultaneous connections.

Do you have an idea why a single thread is unable to handle more connections?
Do you suggest that Falcon is run multi-process, my finding suggest that?
Do you have any suggestions for other benchmarks?

**Comment by ioquatix** [comment]
What is your open file limit?

What is your operating system?

**Comment by edzhelyov** [comment]
MacOS: Catalina
```
ulimit -a
core file size          (blocks, -c) 0
data seg size           (kbytes, -d) unlimited
file size               (blocks, -f) unlimited
max locked memory       (kbytes, -l) unlimited
max memory size         (kbytes, -m) unlimited
open files                      (-n) 20480
pipe size            (512 bytes, -p) 1
stack size              (kbytes, -s) 8192
cpu time               (seconds, -t) unlimited
max user processes              (-u) 2784
virtual memory          (kbytes, -v) unlimited
```
```
sysctl -a | grep ^kern.maxfile
kern.maxfiles: 34576
kern.maxfilesperproc: 24576
```

I'm testing against Nodejs and Crystal simple HTTP servers and they return successful responses, so it's not the file limits.

I'm testing with:
```
ab -c 10000 -n 10000 http://localhost:3000/
```
Where I get `apr_socket_recv: Connection reset by peer (54)` when testing with 1 process and success with 4 processes.

The custom nodeJS script is:
```javascript
const net = require('net')
const http = require('http')

sockets = []
responses = []
errors = []

CONN = 10000

var options = {
  host: 'localhost',
  port: 3000,
  path: '/'
};

callback = function(response) {
  var str = '';

  response.on('data', function (chunk) {
    str += chunk;
  });

  response.on('end', function () {
    responses.push(str)
  });

  response.on('error', function(error) {
    errors.push(error)
  })
}

for(i = 0; i < CONN; i++) {
  r = http.request({host: 'localhost', port: 3000, path: '/'}, callback)

  sockets.push(r)
}
```
Where I manually copy the script above in the node REPL and wait to establish connections and then make requests with:
```javascript
for(i = 0; i < CONN; i++) {
  sockets[i].end()
}
```

**Reply by ioquatix** [reply]
I have personally had a lot of problems running stress tests and benchmarks on macOS network stack. In particular, I noticed that making lots of sequential connections can overload the number of available port mappings. Can you try it on Linux to see if you experience the same issues?

**Comment by edzhelyov** [comment]
I've tested on a Linux machine and the results are much better. With 2 processes Falcon is able to keep up without problems with 10,000 connections. With 1 process with a single thread, it handles around 9,500 connections.

**Comment by edzhelyov** [comment]
Do you have anything like Go's channels in Async, so that two fibers can communicate?

**Reply by ioquatix** [reply]
Yes, `Async::Queue` in the same process, and we are working on a `msgpack` queue for multi-process communication in a separate gem.

**Comment by edzhelyov** [comment]
I'm testing ActiveRecord and "pg" gem and I get very confusing results. With the following sample:
```ruby
require 'async'
require 'pg'
require 'active_record'

ActiveRecord::Base.establish_connection(
  adapter: 'postgresql',
  username: 'receiptbank',
  database: 'revapp_development',
)

class Document < ActiveRecord::Base
end

def console(msg)
  puts "#{Thread.current.object_id}: #{msg}"
end

Async do
  Async do
    id = 1
    loop do
      sleep 1
      console id
      id += 1
    end
  end

  Async do
    console 'pg sleep 5 start'
    Document.connection.execute 'SELECT * FROM pg_sleep(5)'
    console 'pg sleep 5 end'
  end

  Async do
    console 'pg sleep 3 start'
    Document.connection.execute 'SELECT * FROM pg_sleep(3)'
    console 'pg sleep 3 end'
  end
end
```
When I ran it I get the following output:
```shell
500: pg sleep 5 start
500: pg sleep 3 start
500: 1
500: 2
500: 3
500: pg sleep 3 end
500: 4
500: 5
500: pg sleep 5 end
500: 6
```
And from that output it seems that the ActiveRecord connection is not blocking the other Fibers.

But if I put the code as Sinatra app and run it with Falcon:
```ruby
require 'sinatra'
require 'pg'
require 'active_record'

ActiveRecord::Base.establish_connection(
  adapter: 'postgresql',
  username: 'receiptbank',
  database: 'revapp_development',
)

class Document < ActiveRecord::Base
end

def console(msg)
  puts "#{Thread.current.object_id}: #{msg}"
end

get '/' do
  console 'req start'
  Document.connection.execute 'SELECT * FROM pg_sleep(2)'
  console 'req end'
end
```
I ran it with only one Falcon instance: `falcon -b http://localhost:3000 -c falcon-db.ru -n 1` so I expect one Thread to serve the requests.
And test wth `ab -n 5 -c 5 http://localhost:3000/` I get:
```shell
3080: req start
3080: req end
::1 - - [06/Dec/2021:23:09:21 +0200] "GET / HTTP/1.0" 200 - 2.0825
3080: req start
3080: req start
3080: req start
3080: req start
3080: req end
::1 - - [06/Dec/2021:23:09:23 +0200] "GET / HTTP/1.0" 200 - 2.0762
3080: req end
::1 - - [06/Dec/2021:23:09:25 +0200] "GET / HTTP/1.0" 200 - 4.1462
3080: req end
::1 - - [06/Dec/2021:23:09:27 +0200] "GET / HTTP/1.0" 200 - 6.2215
3080: req end
::1 - - [06/Dec/2021:23:09:29 +0200] "GET / HTTP/1.0" 200 - 8.2952
```
The requests are serialized instead of processed in concurrenctly. Do you have any explanations why there is a difference, my assumption is that pg and ActiveRecord should block the fibers.

**Reply by ioquatix** [reply]
In order to log AR connection contention, you'd need to log:

1. Before taking the lock.
2. After entering the mutex.
3. Before exiting the mutex.
4. After releasing the lock.

You should see that 2 and 3 are not interleaved with any other request, but it appears you are only logging 1 and 4 which can be interleaved. Can you check it?

**Reply by edzhelyov** [reply]
I'm not sure to which mutex you are referring. Can you be more specific?

**Reply by ioquatix** [reply]
The per-connection mutex in AR.

**Reply by edzhelyov** [reply]
Sorry, I feel a bit lost, most likely my understanding of AR, Async, and concurrent systems are lacking. Can you help me a bit?
Do you expect calls to AR to block the fibers?
